### DISCUSSION TITLE
Collect .so files in hook

### DISCUSSION THREAD

**Discussion by arossert** [pyinstaller/pyinstaller#6589] [Help]
I'm writing a hook to my module and I also want to include `*.so` files as `datas`, I am using `datas = collect_data_files("my module")` but the `*.so` files are not included.

After looking at the code I noticed this:
```
    # Including py files means don't exclude them. This pattern will search any directories for containing files, so
    # do not modify ``excludes_len``.
    if not include_py_files:
        excludes += ['**/*' + s for s in compat.ALL_SUFFIXES]
```
One of the suffixes in `compat.ALL_SUFFIXES` is `.so` and it seems that the only option is to use `include_py_files=True`.

What I'm currently doing is:
```
excludes = ["**/*" + s for s in set(compat.ALL_SUFFIXES + [".pyo"]) - {".so"}]
datas = collect_data_files("my module", include_py_files=True, excludes=excludes)
```

Is there any other way?

**Comment by bwoodsend** [comment]
You're right I'm afraid. It's a hole in its capabilities. I'm not entirely sure what the API should be but it does need some way of separating plain .py files from extension modules.

**Reply by arossert** [reply]
Maybe if `excludes` param is used it should override the default excludes that are implied in the function.

**Reply by bwoodsend** [reply]
That would mean that excluding a particular subdirectory would include .py files which is certainly not what I'd expect.

**Comment by rokm** [comment]
I'd also like to point out that unless you have a good reason to put .so files under `datas` (e.g., trying to work-around path rewriting on macOS), shared libraries should preferably be put under `binaries`  as opposed to `datas` (otherwise they are not included in binary analysis on any OS, nor in macOS-specific magic like path rewriting).

**Reply by arossert** [reply]
This is a specific case when I need the .so file as a binary and I do not need any magic around it 😃 
I also have .dll files that are collected without any issues.
